crates/uv/src/commands/project/mod.rs

```diff
@@ -1632,27 +1632,49 @@ pub(crate) async fn resolve_names(
     Ok(requirements)
 }
 
+#[derive(Debug, Clone)]
+pub(crate) enum EnvironmentPreferences<'lock> {
+    /// The preferences should be extracted from a lockfile.
+    Lock {
+        lock: &'lock Lock,
+        install_path: &'lock Path,
+    },
+    /// The preferences will be provided directly as [`Preference`] entries.
+    Entries(Vec<Preference>),
+}
+
 #[derive(Debug, Clone)]
 pub(crate) struct EnvironmentSpecification<'lock> {
     /// The requirements to include in the environment.
     requirements: RequirementsSpecification,
-    /// The lockfile from which to extract preferences, along with the install path.
-    lock: Option<(&'lock Lock, &'lock Path)>,
+    /// The preferences to respect when resolving.
+    preferences: Option<EnvironmentPreferences<'lock>>,
 }
 
 impl From<RequirementsSpecification> for EnvironmentSpecification<'_> {
     fn from(requirements: RequirementsSpecification) -> Self {
         Self {
             requirements,
-            lock: None,
+            preferences: None,
         }
     }
 }
 
 impl<'lock> EnvironmentSpecification<'lock> {
     #[must_use]
-    pub(crate) fn with_lock(self, lock: Option<(&'lock Lock, &'lock Path)>) -> Self {
-        Self { lock, ..self }
+    pub(crate) fn with_lock(self, lock: &'lock Lock, install_path: &'lock Path) -> Self {
+        Self {
+            preferences: Some(EnvironmentPreferences::Lock { lock, install_path }),
+            ..self
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn with_preferences(self, entries: Vec<Preference>) -> Self {
+        Self {
+            preferences: Some(EnvironmentPreferences::Entries(entries)),
+            ..self
+        }
```

Comment:
nit: Should those two be a single method? Otherwise it's hard to spot that one overwrites the other.

Reply:
Good call.